crates/muvm/src/bin/muvm.rs

```diff
@@ -383,58 +362,59 @@ fn main() -> Result<ExitCode> {
-    let muvm_server_path = find_muvm_exec("muvm-server")?;
 
-    let mut muvm_guest_args: Vec<String> = vec![
+    let display = env::var("DISPLAY").ok();
+    let guest_config = GuestConfiguration {
+        command: Launch {
+            cookie,
+            command,
+            command_args,
+            env: HashMap::new(),
+            vsock_port: 0,
+            tty: false,
+            privileged: false,
+        },
+        server_port: options.server_port,
+        username,
+        uid: getuid().as_raw(),
+        gid: getgid().as_raw(),
+        host_display: display,
+        server_cookie: cookie,
+        merged_rootfs: options.merged_rootfs,
+    };
+    let mut muvm_config_file = NamedTempFile::new()
+        .context("Failed to create a temporary file to store the muvm guest config")?;
+    write!(
+        muvm_config_file,
+        "{}",
+        serde_json::to_string(&guest_config)
+            .context("Failed to transform GuestConfiguration into a JSON string")?
+    )
+    .context("Failed to write to temporary config file")?;
+
+    let muvm_config_path = muvm_config_file
+        .path()
+        .to_str()
+        .context("Temporary directory path contains invalid UTF-8")?
+        .to_string();
+    let muvm_guest_args = vec![
         muvm_guest_path
             .to_str()
             .context("Failed to process `muvm-guest` path as it contains invalid UTF-8")?
-            .to_owned(),
-        username,
-        format!("{uid}", uid = getuid().as_raw()),
-        format!("{gid}", gid = getgid().as_raw()),
-        muvm_server_path
-            .to_str()
-            .context("Failed to process `muvm-server` path as it contains invalid UTF-8")?
-            .to_owned(),
-        command
-            .to_str()
-            .context("Failed to process command as it contains invalid UTF-8")?
             .to_string(),
+        muvm_config_path,
     ];
-    for arg in command_args {
-        muvm_guest_args.push(arg);
-    }
 
-    env.insert(
-        "MUVM_SERVER_PORT".to_owned(),
-        options.server_port.to_string(),
-    );
-    env.insert(
-        "MUVM_ROOT_SERVER_PORT".to_owned(),
-        options.root_server_port.to_string(),
-    );
-    env.insert("MUVM_SERVER_COOKIE".to_owned(), cookie.to_string());
-
-    if !options.sommelier {
-        let display =
-            env::var("DISPLAY").context("X11 forwarding requested but DISPLAY is unset")?;
-        env.insert("HOST_DISPLAY".to_string(), display);
-
-        // And forward XAUTHORITY. This will be modified to fix the
-        // display name in muvm-guest.
-        if let Ok(xauthority) = env::var("XAUTHORITY") {
-            env.insert("XAUTHORITY".to_string(), xauthority);
-        }
+    // And forward XAUTHORITY. This will be modified to fix the
+    // display name in muvm-guest.
+    if let Ok(xauthority) = env::var("XAUTHORITY") {
+        env.insert("XAUTHORITY".to_string(), xauthority);
     }
 
     let mut krun_config = KrunConfig {
```

Comment:
```suggestion
    let krun_config = KrunBaseConfig {
        config: KrunConfig {
            args: muvm_guest_args,
            envs: env
                .into_iter()
                .map(|(key, value)| format!("{key}={value}"))
                .collect(),
        },
    };
```

(Sorry, GitHub's code suggestions UI is considerably bugged now...)